src/Network/Transport/Chan.hs

```diff
@@ -29,72 +67,212 @@ data TransportState = State { _channels         :: Map EndPointAddress (Chan Eve
+                atomically $ writeTVar (localConnectionState con) LocalConnectionClosed
+    }
+
+-- | Create a new end point.
+apiNewEndPoint :: TVar TransportState -> IO (Either (TransportError NewEndPointErrorCode) EndPoint)
+apiNewEndPoint state = handle (return . Left) $ atomically $ do
+  chan <- newTChan
+  (lep,addr) <- withValidTransportState state NewEndPointFailed $ \vst -> do
+    lepState <- newTVar $ LocalEndPointValid $ ValidLocalEndPointState
+      { _nextConnectionId = 1
+      , _connections = Map.empty
+      , _multigroups = Map.empty
+      }
+    let r = nextLocalEndPointId ^: (+ 1) $ vst
+        addr = EndPointAddress . BSC.pack . show $ r ^. nextLocalEndPointId
+        lep = LocalEndPoint
+          { localEndPointAddress = addr
+          , localEndPointChannel = chan
+          , localEndPointState = lepState
+          }
+    writeTVar state (TransportValid $! localEndPointAt addr ^= Just lep $ r)
+    return (lep, addr)
+  return $ Right $ EndPoint
+    { receive       = atomically $ do
+        result <- tryReadTChan chan
+        case result of
+          Nothing -> do st <- readTVar (localEndPointState lep)
+                        case st of
+                          LocalEndPointClosed -> throwSTM (userError "Channel is closed.")
+                          LocalEndPointValid{} -> retry
+          Just x -> return x
+    , address       = addr
+    , connect       = apiConnect addr state
+    , closeEndPoint = apiCloseEndPoint state addr
+    , newMulticastGroup     = apiNewMulticastGroup state addr
+    , resolveMulticastGroup = apiResolveMulticastGroup state addr
+    }
+
+apiCloseEndPoint :: TVar TransportState -> EndPointAddress -> IO ()
+apiCloseEndPoint state addr = atomically $ whenValidTransportState state $ \vst ->
+    forM_ (vst ^. localEndPointAt addr) $ \lep -> do
+      old <- swapTVar (localEndPointState lep) LocalEndPointClosed
+      case old of
+        LocalEndPointClosed -> return ()
+        LocalEndPointValid lepvst -> do
+          forM_ (Map.elems (lepvst ^. connections)) $ \lconn -> do
+            st <- swapTVar (localConnectionState lconn) LocalConnectionClosed
+            case st of
+              LocalConnectionClosed -> return ()
+              LocalConnectionFailed -> return ()
+              _ -> forM_ (vst ^. localEndPointAt (localConnectionRemoteAddress lconn)) $ \thep ->
+                     whenValidLocalEndPointState thep $ \_ -> do
+                        writeTChan (localEndPointChannel thep)
+                                   (ConnectionClosed (localConnectionId lconn))
+          writeTChan (localEndPointChannel lep) EndPointClosed
+          writeTVar  (localEndPointState lep)    LocalEndPointClosed
+      writeTVar state (TransportValid $! (localEndPoints ^: Map.delete addr) vst)
+
+apiBreakConnection :: TVar TransportState -> EndPointAddress -> EndPointAddress -> String -> STM ()
```

Comment:
Add comment explaining that this function is to tear down a connection if a `send` fails.